### PR TITLE
Implementation suggestion for "Prefer words" (#5629)

### PR DESCRIPTION
**Pull request by markprins**
PR Is based my own wishes and needs for a "Prefer words" option as is discussed in #5629 

As much as possible I tried to follow current guidelines and general best-practices. I also tried to contact dev on IRC to get some pointers/suggestions on how to create more specific unit-tests to help.

Unfortunately I was not able to get any specific help, so I hope that by using this pull-request I can get some help for merging.

To layout the implemented changes:

- Added hasPreferedWords to show_name_helper
- hasPreferedWords goes through the fetched results and finds out if any of the prefered words are in the list
- If no prefered words exists, it does nothing
- If prefered words exists, it will return a sorted list with the most prefered item at the top
- Also has unit tests available which should capture any issues
- Implements a prefered words option in the settings
- Implements a DB migration (although Im not really sure I did it correct)

(be gentle and/or poke me on IRC, though time-zones may make this tricky)

### CODE REVIEW

**Comment by WebSpider** [comment]
Please point this merge request to the develop branch

File: sickbeard/search.py [hunk 1]
```diff
@@ -418,6 +421,10 @@ def searchForNeededEpisodes():
             "No NZB/Torrent providers found or enabled in the sickchill config for daily searches. Please check your settings.",
             logger.INFO)
 
+
+    logger.log("STOPPING RIGHT HERE")
+    import pdb; pdb.set_trace()
+
```
Comment (markprins) [inline sickbeard/search.py]: Noticed I left debugging/logging statements. Will update later

**Comment by miigotu** [comment]
Is this ready for a merge?

**Comment by markprins** [comment]
I've updated the PR with the latest changes from develop.

Have been running my version locally without problems (works like a charm). It should be ready for merge, but as stated in the pull-request, I'm not entirely sure that I did all the db migrations correctly and if it is up-to-spec.

File: sickbeard/databases/mainDB.py [hunk 2]
```diff
@@ -1155,10 +1155,26 @@ def execute(self):
         backupDatabase(self.get_db_version())
         self.add_column('tv_shows', 'sub_use_sr_metadata', "NUMERIC", "0")
 
+class AddPreferWords(UseSickChillMetadataForSubtitle):
+    """ Adding column rls_prefer_words to tv_shows """
 
-class ResetDBVersion(UseSickChillMetadataForSubtitle):
+    def test(self):
+        return self.get_db_version() >= 45
```
Comment (miigotu) [inline sickbeard/databases/mainDB.py]:
```python
        return self.has_column("tv_shows", "rls_prefer_words")

```

File: sickbeard/databases/mainDB.py [hunk 3]
```diff
@@ -34,7 +34,7 @@
 from sickchill.helper.encoding import ek
 
 MIN_DB_VERSION = 9  # oldest db version we support migrating from
-MAX_DB_VERSION = 44
+MAX_DB_VERSION = 45
```
Comment (miigotu) [inline sickbeard/databases/mainDB.py]: Leave this at 44

File: sickbeard/databases/mainDB.py [hunk 4]
```diff
@@ -1155,10 +1155,26 @@ def execute(self):
...
+        return self.get_db_version() >= 45
+
+    def execute(self):
+        backupDatabase(self.get_db_version())
+
+        logger.log("Adding column rls_prefer_words to tvshows")
+        if not self.has_column("tv_shows", "rls_prefer_words"):
+            self.add_column("tv_shows", "rls_prefer_words", "TEXT", "")
+
+        self.increment_db_version()
```
Comment (miigotu) [inline sickbeard/databases/mainDB.py]:
```python
self.inc_minor_version()
```

File: sickbeard/databases/mainDB.py [hunk 5]
```diff
@@ -1155,10 +1155,26 @@ def execute(self):
...
+
+
+class ResetDBVersion(AddPreferWords):
```
Comment (miigotu) [inline sickbeard/databases/mainDB.py]: Remove this reset class. db should be 44.1 after this pr

**Review by miigotu — CHANGES_REQUESTED** [review]
All looks good except these few things with the DB version. We want to use minor db versions for smaller changes like this for now

**Comment by markprins** [comment]
> All looks good except these few things with the DB version. We want to use minor db versions for smaller changes like this for now

I have incorporated the suggested changes as you described.

As I was running for a few weeks with my branch, I was wondering how I would go about "fixing" my database so it _thinks_ it is the same as would be when it migrates to a new minor version (so I don't need to revert back to my previous db from a month or so back and re-scan everything). Is there an easy fix for that? (I'd think something like set the version manually or something like that).

**Comment by miigotu** [comment]
@markprins Use sqlite3 or another sqlite db editor and change db_version.db_version to 44 and db_version.db_minor_version to 1

If using sqlite3:
```bash
sqlite3 /path/to/sickbeard.db
```
```sql
UPDATE db_version SET db_verion=44, db_minor_version=3;
```

**Review by miigotu — APPROVED** [review]

**Comment by miigotu** [comment]
@markprins sorry I used your branch for a sandbox to fix the windows build tests. Hopefully it is working now.

**Pull request merged**